Update loader documentation. (#685)
KhronosGroup/Vulkan-Loader: docs/LoaderApplicationInterface.md

## Overview

This is the Application-centric view of working with the Vulkan loader.
For the complete overview of all sections of the loader, please refer 
to the [LoaderInterfaceArchitecture.md](LoaderInterfaceArchitecture.md) file.

## Interfacing with Vulkan Functions

There are several ways you can interface with Vulkan functions through the
loader:

### Vulkan Direct Exports

The loader library on Windows, Linux, Android, and macOS will export all core
Vulkan and all appropriate Window System Interface (WSI) extensions.
This is done to make it simpler to get started with Vulkan development.
When an application links directly to the loader library in this way, the
Vulkan calls are simple *trampoline* functions that jump to the appropriate
dispatch table entry for the object they are given.

### Directly Linking to the Loader

#### Dynamic Linking

The loader is ordinarily distributed as a dynamic library (.dll on Windows or
.so on Linux or .dylib on macOS) which gets installed to the system path
for dynamic libraries.
Linking to the dynamic library is generally the preferred method of linking to
the loader, as doing so allows the loader to be updated for bug fixes and

> improvements.

Author: We used to, so this is really old.  I'll still remove it.

Furthermore, the dynamic library is generally installed to Windows
systems as part of driver installation and is generally provided on Linux
through the system package manager.
This means that applications can usually expect a copy of the loader to be
present on a system. If applications want to be completely sure that a loader
is present, they can include a loader or runtime installer with their

> application.

Author: Hmmm.... Yeah, that seems out of date.  I'll remove it.

> application.

Author: As a note, Steam does ship with it's own loader.  But I'm not sure if we want to encourage this behavior.  It is a way for an application to verify that it has at least a loader version it wants (i.e. 1.2).  Do you think I should still remove this or modify it to say that this should only be done in rare cases or some such boiler plate?

> application.

Reviewer: I'm not sure, it certainly is true that there are cases where 'shipping a loader' isn't bad. With steam & app stores, they usually *require* automatic updates, which negates the main issue of the loader being out of date. Its when there isn't a good update mechanism that make me worried, such as static linking or a standalone game executable.

versions with the same major number (e.g. 1.0 and 1.1).
On Windows, the loader library encodes the ABI version in its name such that
multiple ABI incompatible versions of the loader can peacefully coexist on a
given system.
The Vulkan loader library file name is `vulkan-<ABI version>.dll`.
For example, for Vulkan version 1.X on Windows the library filename is
`vulkan-1.dll`.
And this library file can typically be found in the `windows\system32`
directory (on 64-bit Windows installs, the 32-bit version of the loader with
the same name can be found in the `windows\sysWOW64` directory).

For Linux and macOS, shared libraries are versioned based on a suffix.
Thus, the ABI number is not encoded in the base of the library filename as on
Windows.
On Linux an application wanting to link to the latest Vulkan ABI version would
just link to the name vulkan (`libvulkan.so`).  A specific Vulkan ABI version
can also be linked to by applications (e.g. `libvulkan.so.1`).
On macOS, the libraries are `libvulkan.dylib` and `libvulkan.1.dylib`.

## Application Layer Usage

Applications desiring Vulkan functionality beyond what the core API offers may

> use various layers or extensions.

Author: I don't like your new version either.  How about instead:

`Applications desiring Vulkan functionality beyond what Vulkan implementations
on their system already expose, may use various layers to augment the API.`

> use various layers or extensions.

Reviewer: A lot better, though I have one nit: drop the comma.

A layer cannot introduce new Vulkan core API entry-points to an application
that are not exposed in Vulkan.h.
However, layers may offer extensions that introduce new Vulkan commands that

> can be queried through the extension interface.

Author: How about this instead?

```
However, layers may offer implementations of extensions that introduce
additional entry-points beyond what is available without those layers.
These additional extension entry-points can be queried through the Vulkan
extension interface.
```

> can be queried through the extension interface.

Reviewer: Looks great.

A common use of layers is for API validation which can be enabled by loading
the layer during application development, but not loading the layer for
application release.

> This eliminates the overhead of validating the application's usage of the API,
> something that wasn't available on some previous graphics APIs.

Reviewer: ```suggestion
This allows easy control of the overhead from validating the application's usage of the API,
which wasn't always possible in previous graphics APIs.
```

Author: How about:

```
This allows easy control of the overhead resulting from enabling validation of
the application's usage of the API, which wasn't always possible in previous
graphics APIs.
```

Reviewer: Looks great.

To find out what layers are available to your application, use
`vkEnumerateInstanceLayerProperties`.
This will report all layers that have been discovered by the loader.
The loader looks in various locations to find layers on the system.
For more information see the
[Layer discovery](LoaderLayerInterface.md#layer-discovery)
section in the
[LoaderLayerInterface.md document](LoaderLayerInterface.md) document.

> To enable a layer, or layers, simply pass the name of the layers you wish to

Reviewer: ```suggestion
To enable specific layers, simply pass the names of the layers you wish to
```

enable in the `ppEnabledLayerNames` field of the `VkInstanceCreateInfo` during
a call to `vkCreateInstance`.
Once done, the layers you have enabled will be active for all Vulkan functions
using the created `VkInstance`, and any of its child objects.

**NOTE:** Layer ordering is important in several cases since some layers
interact with each other.
Be careful when enabling layers as this may be the case.
See the [Overall Layer Ordering](#overall-layer-ordering) section for more
information.

The following code section shows how you would go about enabling the
VK_LAYER_KHRONOS_validation layer.

```

> char *instance_validation_layers[] = {

Reviewer: ```suggestion
char *instance_layers[] = {
```
There are more than validation layers.

    "VK_LAYER_KHRONOS_validation"
};
const VkApplicationInfo app = {
    .sType = VK_STRUCTURE_TYPE_APPLICATION_INFO,
    .pNext = NULL,
    .pApplicationName = "TEST_APP",
    .applicationVersion = 0,
    .pEngineName = "TEST_ENGINE",
    .engineVersion = 0,
    .apiVersion = VK_API_VERSION_1_0,
};
VkInstanceCreateInfo inst_info = {
    .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
    .pNext = NULL,
    .pApplicationInfo = &app,
    .enabledLayerCount = 1,

>     .ppEnabledLayerNames = (const char *const *)instance_validation_layers,

Reviewer: ```suggestion
    .ppEnabledLayerNames = (const char *const *)instance_layers,
```
Same as above

    .enabledExtensionCount = 0,
    .ppEnabledExtensionNames = NULL,
};

> err = vkCreateInstance(&inst_info, NULL, &demo->inst);

Reviewer: Possibly check for `VK_ERROR_LAYER_NOT_PRESENT`? Since its a something to think of. Feel free to dismiss, this is not 'demo code' just showcasing how layers are enabled.

```

At `vkCreateInstance` and `vkCreateDevice`, the loader constructs call chains
that include the application specified (enabled) layers.
Order is important in the `ppEnabledLayerNames` array; array element 0 is the
topmost (closest to the application) layer inserted in the chain and the last
array element is closest to the driver.
See the [Overall Layer Ordering](#overall-layer-ordering) section for more
information on layer ordering.

**NOTE:** *Device Layers Are Now Deprecated*
> `vkCreateDevice` originally was able to select layers in a similar manner to
`vkCreateInstance`.
> This lead to the concept of "instance layers" and "device layers".
> It was decided by Khronos to deprecate the "device layer" functionality and
> only consider "instance layers".
> Therefore, `vkCreateDevice` will use the layers specified at
`vkCreateInstance`.
> Because of this, the following items have been deprecated:
> * `VkDeviceCreateInfo` fields:
>   * `ppEnabledLayerNames`
>   * `enabledLayerCount`
> * The `vkEnumerateDeviceLayerProperties` function

### Meta-Layers

> Meta-layers are layers which really just contain an ordered list of other
> layers.

Reviewer: ```suggestion
Meta-layers are layers which contain an ordered list of other layers to enable.
```

This is to allow grouping layers together in a specified order so that they can
interact properly.

> Originally, this was designed to group together the individual Vulkan
> Validation layers in the proper order to avoid conflicts.
> While the Vulkan Validation layer no longer uses this, VkConfig does use this
> to group layers together based on user's desires.

Reviewer: ```suggestion
Originally, this was used to group together the individual Vulkan Validation layers in the proper order
to avoid conflicts. It was necessary because instead of a single Validation layer, validation was split into
multiple component layers. The new `VK_LAYER_KHRONOS_validation` layer pulled everything
into a single layer, dropping the need for meta layers. While not necessary for validation anymore,
VkConfig does use meta layers to group layers together based on user's preferences.
```

More can be found out about this functionality through both the
[VkConfig documentation](https://github.com/LunarG/VulkanTools/blob/master/vkconfig/README.md)
and the section later on the [Override Layer](#override-layer).

Meta-layers are detailed more in the
[Meta-Layers](LoaderLayerInterface.md#meta-layers) section of the
[LoaderLayerInterface.md](LoaderLayerInterface.md) file in this folder.

### Implicit vs Explicit Layers

![Different Types of Layers](./images/loader_layer_order.png)

Explicit layers are layers which are enabled by an application (e.g. with the
vkCreateInstance function as mentioned previously), or by an environment

> variable.

Reviewer: Technically, layers enabled with an environment variable are implicit layers. They just are *not* enabled all the time. I agree with the intent of this sentence, but not the specifics. (Because to the loader, Implicit layers means layers with X properties in the json manifest. Layers that get enabled with an env-var are use the rules defined by implicit layers, not explicit.
The following sentences discusses 'certain application environments' which are what env-var enabled layers usually are used for.

Author: Good point, really the only thing you can do with explicit layers with environment variables is change their path.  I'll remove the last part of this.

Author: Done, please look over new version.

Author: `Explicit layers are layers which are enabled by an application (e.g. with the
vkCreateInstance function as mentioned previously).`

infotainment system) may have layers which they always want enabled for all
applications that they start.
Other implicit layers may be for all applications started on a given system
(e.g. layers that overlay frames-per-second).
Implicit layers are enabled automatically, whereas explicit layers must be
enabled explicitly.

Implicit layers have an additional requirement over explicit layers in that
they require being able to be disabled by an environmental variable.
This is due to the fact that they are not visible to the application and could
cause issues.
A good principle to keep in mind would be to define both an enable and disable
environment variable so the users can deterministically enable the
functionality.
On Desktop platforms (Windows, Linux, and macOS), these enable/disable settings
are defined in the layer's JSON file.

Discovery of system-installed implicit and explicit layers is described later
in the [Layer discovery](LoaderLayerInterface#layer-discovery)
section in the
[LoaderLayerInterface.md document](LoaderLayerInterface.md) document.
For now, simply know that what distinguishes a layer as implicit or explicit

> is dependent on the Operating system, as shown in the table below.

Reviewer: "Distinguishes a layer as implicit or explicit is dependent on the OS" is a weird choice of words, because what distinguishes the two really depends on *where* the layers are located on an OS more than the OS itself. 
I do see how android doesn't allow implicit layers, so that's justifiable, but otherwise its not super clear.

Author: That is nasty.  Will update.

Author: Done, please look over new version.

<table style="width:100%">
  <tr>
    <th>Operating System</th>
    <th>Implicit Layer Identification</th>
  </tr>
  <tr>
    <td>Windows</td>
    <td>Implicit layers are located in a different Windows registry location
        than explicit layers.</td>
  </tr>
  <tr>
    <td>Linux</td>
    <td>Implicit layers are located in a different directory location than
        explicit layers.</td>
  </tr>
  <tr>
    <td>Android</td>
    <td>There is **No Support For Implicit Layers** on Android.</td>
  </tr>
  <tr>
    <td>macOS</td>
    <td>Implicit layers are located in a different directory location than
        explicit layers.</td>
  </tr>
</table>

#### Override Layer

The "Override Layer" is a special implicit meta-layer created by the 
[VkConfig](https://github.com/LunarG/VulkanTools/blob/master/vkconfig/README.md)
tool and available by default when the tool is running.
Once VkConfig exits, the override layer is removed, and the system should

> return to standard Vulkan behavior.

Reviewer: We should remove the loader code responsible for the 'expiration' because that was a silly idea that doesn't live on in the current vkconfig application.

Author: Kind of orthogonal to this.  I don't think I have any mention of that expiring behavior in this doc, do I?

Reviewer: Yeah, definitely is orthogonal. Very much a low priority task, since its not affecting anyone (that we know of)

SDKs, OS package distributions and Independent Hardware Vendor (IHV) driver
packages.
These details are beyond the scope of this document.
However, the name and versioning of the Vulkan loader library is specified so
an app can link to the correct Vulkan ABI library version.
ABI backwards compatibility is guaranteed for all versions with the same major
number (e.g. 1.0 and 1.1).
On Windows, the loader library encodes the ABI version in its name such that
multiple ABI incompatible versions of the loader can peacefully coexist on a
given system.
The Vulkan loader library file name is `vulkan-<ABI version>.dll`.
For example, for Vulkan version 1.X on Windows the library filename is
`vulkan-1.dll`.
This library file can typically be found in the `windows\system32`
directory (on 64-bit Windows installs, the 32-bit version of the loader with
the same name can be found in the `windows\sysWOW64` directory).

For Linux and macOS, shared libraries are versioned based on a suffix.
Thus, the ABI number is not encoded in the base of the library filename as on
Windows.
On Linux an application wanting to link to the latest Vulkan ABI version would

> just link to the name vulkan (`libvulkan.so`).  A specific Vulkan ABI version

Reviewer 2: In the standard setup for Linux library ABIs, the distinction between `libvulkan.so` and `libvulkan.so.1` on Linux is really about compile-time vs. runtime, rather than latest version vs. a fixed version.

The `.so` symlink is only intended to be used at compile-time: a developer links to libvulkan with something like `cc -lvulkan`, which automatically makes compiler search for `libvulkan.so`.

It would never be safe for an application to load `libvulkan.so` with `dlopen()`, because it could get a Vulkan ABI that is not the one it is expecting, with arbitrary incompatible changes - then it will almost certainly crash. In distributions that separate development libraries from runtime libraries, like Debian and Red Hat, the `.so` symlink is in the `-dev` or `-devel` package and is not installed by end-users, so applications should not assume it will exist.

Author: I applied the changes, I'll resolve this and leave the next one open until you've looked over the changes.

> can also be linked to by applications (e.g. `libvulkan.so.1`).

Reviewer 2: When an application or a larger library links to `libvulkan` with `cc -lvulkan` or similar, the `SONAME` of the library (currently `libvulkan.so.1`) is what gets hard-coded into the binary, so it will always load the same major version that it was compiled against.

If an application or library loads Vulkan-Loader with `dlopen()`, then it should always be using the name that has the ABI suffix, `libvulkan.so.1`, so that it will get the major version that it expects.

Author: That makes sense.  Thanks for clarifying.  I know at the time I wrote this 4 or 5 years ago, I was not aware of that.  I'll update the language.

Author: Changes applied.  Let me know what you think.

layer `VK_LAYER_KHRONOS_validation` on Linux or macOS is as follows:

```
> $ export VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation
```

See the
[Table of Debug Environment Variables](LoaderInterfaceArchitecture.md#table-of-debug-environment-variables)
in the [LoaderInterfaceArchitecture.md document](LoaderInterfaceArchitecture.md)
for more details.

### Overall Layer Ordering

The overall ordering of all layers by the loader based on the above looks
as follows:

![Loader Layer Ordering](./images/loader_layer_order_calls.png)

Ordering may also be important internally to the list of explicit layers.
Some layers may be dependent on other behavior being implemented before
or after the loader calls it.
For example: An overlay layer may want to use `VK_LAYER_KHRONOS_validation`
to verify that the overlay layer is behaving appropriately.
This requires putting the overlay layer closer to the application so that the
validation layer can intercept any Vulkan API calls the overlay layer needs to
make to function.

## Application Usage of Extensions

Extensions are optional functionality provided by a layer, the loader or an

> implementation.

Author: https://www.youtube.com/watch?v=8Gv0H-vPoDc

targets, but others are particular to a given execution environment (and
loader).
This desktop loader (currently targeting Windows, Linux, and macOS) only
enables and directly exports those WSI extensions that are appropriate to the
current environment.
For the most part, the selection is done in the loader using compile-time
preprocessor flags.
All versions of the desktop loader currently expose at least the following WSI
extension support:
- VK_KHR_surface
- VK_KHR_swapchain
- VK_KHR_display

In addition, each of the following OS targets for the loader support target-
specific extensions:

| Windowing System | Extensions available |
|----------------|--------------------|
| Windows  | VK_KHR_win32_surface |
| Linux (Wayland) | VK_KHR_wayland_surface |
| Linux (X11) |  VK_KHR_xcb_surface and VK_KHR_xlib_surface |
| macOS (MoltenVK) | VK_MVK_macos_surface |
| QNX (Screen) | VK_QNX_screen_surface |

It is important to understand that while the loader may support the various
entry-points for these extensions, there is a handshake required to actually
use them:
* At least one physical device must support the extension(s)

> * The application must select such a physical device

Reviewer: ```suggestion
* The application must use such a physical device when creating a logical device.
```

* The application must request the extension(s) be enabled while creating the
instance or logical device (this depends on whether or not the given extension
works with an instance or a device)

> * The instance and/or logical device creation must succeed

Reviewer: ```suggestion
```
(Redundant sentence if you ask me)

Only then can the WSI extension be properly used in a Vulkan program.

### Unknown Extensions

With the ability to expand Vulkan so easily, extensions will be created that
the loader knows nothing about.
If the extension is a device extension, the loader will pass the unknown
entry-point down the device call chain ending with the appropriate
implementation entry-points.
The same thing will happen if the extension is an instance extension which
takes a physical device parameter as its first component.
However, for all other instance extensions the loader will fail to load it.

*But why doesn't the loader support unknown instance extensions?*
<br/>
Let's look again at the instance call chain:

![Instance call chain](./images/loader_instance_chain.png)

Notice that for a normal instance function call, the loader has to handle
passing along the function call to the available implementations.
If the loader has no idea of the parameters or return value of the instance
call, it can't properly pass information along to the implementations.
There may be ways to do this, which will be explored in the future.

> However, for now, this loader does not support instance extensions which don't
> take a physical device as their first parameter.

Reviewer: ```suggestion
However, for now, the loader does not support instance extensions which don't expose entry points that
take a physical device as their first parameter.
```

Because the device call-chain does not normally pass through the loader
*terminator*, this is not a problem for device extensions.
Additionally, since a physical device is associated with one implementation, we
can use a generic *terminator* pointing to one implementation.
This is because both of these extensions terminate directly in the
implementation they are associated with.

*Is this a big problem?*
<br/>
No!
Most extension functionality only affects either a physical or logical device
and not an instance.
Thus, the overwhelming majority of extensions should be supported with direct
loader support.

### Filtering Out Unknown Instance Extension Names

> In some cases, an implementation may support instance extensions that the loader
> does not.

Reviewer: ```suggestion
In some cases, an implementation may support instance extensions that are not supported by the loader.
```

For the above reasons, the loader will filter out the names of these unknown
instance extensions when an application calls
`vkEnumerateInstanceExtensionProperties`.

> Additionally, this behavior will cause the loader to throw an error during
> `vkCreateInstance` if the application still attempts to use one of these

Reviewer: ```suggestion
Additionally, this behavior will cause the loader to emit an error during
`vkCreateInstance` if the application still attempts to use one of these
```

### ABI Versioning

The Vulkan loader library will be distributed in various ways including Vulkan
SDKs, OS package distributions and Independent Hardware Vendor (IHV) driver
packages.
These details are beyond the scope of this document.
However, the name and versioning of the Vulkan loader library is specified so
an app can link to the correct Vulkan ABI library version.
ABI backwards compatibility is guaranteed for all versions with the same major
number (e.g. 1.0 and 1.1).
On Windows, the loader library encodes the ABI version in its name such that
multiple ABI incompatible versions of the loader can peacefully coexist on a
given system.
The Vulkan loader library file name is `vulkan-<ABI version>.dll`.
For example, for Vulkan version 1.X on Windows the library filename is
`vulkan-1.dll`.
This library file can typically be found in the `windows\system32`
directory (on 64-bit Windows installs, the 32-bit version of the loader with
the same name can be found in the `windows\sysWOW64` directory).

For Linux and macOS, shared libraries are versioned based on a suffix.
Thus, the ABI number is not encoded in the base of the library filename as on
Windows.

> When an application or a larger library links to `libvulkan.so` (with something
> like "cc -lvulkan"), the SONAME of the library (currently `libvulkan.so.1`) is
> what gets hard-coded into the binary, so it will always load the same major
> version that it was compiled against.
> On Linux an application wanting to link to the latest Vulkan ABI version would
> just link to the name vulkan (`libvulkan.so`).

Reviewer 2: Sorry, I hadn't realised you were going to incorporate my comment into the document as-is - I was writing review comments in a conversational style, rather than writing for inclusion in documentation.

I think the current text  has the potential to be misleading, because there are two ways an application can link to the Vulkan library - ordinary compile-time linking, and dynamic loading at runtime - and they have different patterns.

For compile-time linking, like `vkcube` does, what we want (and the only thing that works in most build systems) is for the application to ask for `-lvulkan`, and the compiler and linker to resolve this to a specific SONAME `libvulkan.so.1`.

For dynamic loading, like SDL does, what we want is for the developer to load `libvulkan.so.1` specifically.

How about this, perhaps?

```suggestion
On Linux, applications that have a hard dependency on Vulkan should request
linking to the unversioned name `libvulkan.so` in their build system, for
example by importing the CMake target `Vulkan::Vulkan` or by using the
output of `pkg-config --cflags --libs vulkan` as compiler flags.
As usual for Linux libraries, the compiler and linker will resolve this to
a dependency on the correct versioned SONAME, currently `libvulkan.so.1`.
Linux applications that load Vulkan-Loader dynamically at runtime do not
benefit from this mechanism, and should instead make sure to pass the
versioned name such as `libvulkan.so.1` to `dlopen()`, to ensure that they
load a compatible version.
```

Technically this is also true for most non-Linux ELF systems like FreeBSD and Solaris, but I know Linux is the only one that's officially supported, and developers on non-Linux Unix are probably already familiar with extrapolating what's meant to happen on their platform from what happens on Linux.

I don't know how this works on macOS (I don't develop on Apple systems).

Author: I like your version.  Plus, I ran it by the internal Mac savy people and they say the language works there as well.
